src/UCI.cpp

```diff
@@ -311,8 +311,15 @@ void UCI::loop(const std::string& start) {
           myprintf_so("id name lczero " PROGRAM_VERSION "\nuciok\n");
       }
       else if (token == "setoption")  setoption(is);
-      else if (token == "go")         go(search,bh,is);
-      else if (token == "stop")       search.please_stop();
+      else if (token == "go") {
+          search.please_stop();
+          if (search_thread && search_thread->joinable()) search_thread->join();
+          search_thread = go(search, bh, is);
+      }
+      else if (token == "stop") {
+          search.please_stop();
+          if (search_thread && search_thread->joinable()) search_thread->join();
+      }
       else if (token == "perft")      uci_perft(bh, is);
       else if (token == "position")   position(bh, is);
       else if (token == "ucinewgame") ;
```

Comment:
I think the fact this does nothing is a bug.  I think it can cause tree reuse contamination across games making new games play stronger than old games.

Reply:
I think it used to do something, but was commented out because position automatically checks if the move is legal in bh and resets it if not. It was like that before I joined the project though so let me know if I missed something.

Reply:
you are right, I'll take a closer look at the potential contamination scenario, but it definitely shouldn't block this PR.